services/linkchecker/src/main/java/net/geocat/database/linkchecker/service/ServiceDocumentLinkService.java

```diff
@@ -72,11 +72,64 @@ public DatasetDocumentLink create(DatasetMetadataRecord datasetMetadataRecord, O
         result.setFunction(onlineResource.getFunction());
         result.setOperationName(onlineResource.getOperationName());
         result.setRawURL(onlineResource.getRawURL());
-        result.setProtocol(onlineResource.getProtocol());
+
+        String protocolFromUrl = inferProtocolFromUrl(onlineResource.getRawURL());
+
+        if ((onlineResource.getProtocol() == null) && (protocolFromUrl != null)) {
+            // If no protocol defined, try to infer the protocol from the URL
+            result.setProtocol(protocolFromUrl);
+        } else {
+            result.setProtocol(onlineResource.getProtocol());
+
+            // TODO: review
+            // Check if protocol inferred is not the same type to the one defined in the protocol field and update it based on the URL inferred protocol
+            if (protocolFromUrl != null) {
+                boolean isDownloadProtocol = ServiceDocumentLink.validDownloadProtocols.contains(onlineResource.getProtocol().toLowerCase());
+                boolean isDownloadUrlProtocol = ServiceDocumentLink.validDownloadProtocols.contains(protocolFromUrl.toLowerCase());
+                boolean isViewProtocol = ServiceDocumentLink.validViewProtocols.contains(onlineResource.getProtocol().toLowerCase());
+                boolean isViewUrlProtocol = ServiceDocumentLink.validViewProtocols.contains(protocolFromUrl);
+
+                if (isDownloadProtocol) {
```

Comment:
This is a bit difficult to read.

// if the XML document's protocol isn't compatible with the actual URL 
// then use the the inferred URL protocol.
// Example;
//     xml protocol is WMS (view)
//     but, url is "...?service=WFS" (inferred url protocol is download and not view)
//     then, set the protocol to Download (ignore the XML)

Reply:
thanks @davidblasby, updated the code